tests/pocket_test.cpp

```diff
@@ -0,0 +1,832 @@
+#include "ammo.h"
+#include "item.h"
+#include "item_pocket.h"
+
+#include "catch/catch.hpp"
+
+// Pocket Tests
+// ------------
+//
+// These tests are focused on the `item_pocket` and `pocket_data` classes, and how they implement
+// the "pocket_data" section of JSON item definitions (as documented in doc/JSON_FLAGS.md).
+//
+// To run all tests in this file:
+//
+//      tests/cata_test [pocket]
+//
+// Other tags used include: [holster], [magazine], [airtight], [watertight], [rigid], and more.
+//
+// Most of the items used for testing here are defined in data/mods/TEST_DATA/items.json, to have
+// a predictable set of data unaffected by in-game balance changes.
+
+
+// TODO: Add tests focusing on these JSON pocket_data fields:
+// -  pocket_type
+// -  spoil_multiplier
+// -  weight_multiplier
+// -  magazine_well
+// -  fire_protection
+// -  open_container
+// -  sealed_data
+// -  moves
+
+// FIXME: The failure messaging from these helper functions is pretty terrible, due to the multiple
+// CHECKs in each function. Capturing it.tname() helps only slightly. The `rate_can.value()` is an
+// integer, so it is difficult to see when a failure reason is returned instead of success.
+
+// Expect pocket.can_contain( it ) to be successful (contain_code::SUCCESS)
+static void expect_can_contain( item_pocket &pocket, item &it )
+{
+    CAPTURE( it.tname() );
+    ret_val<item_pocket::contain_code> rate_can = pocket.can_contain( it );
+    CHECK( rate_can.success() );
+    CHECK( rate_can.str().empty() );
+    CHECK( rate_can.value() == item_pocket::contain_code::SUCCESS );
+}
+
+// Expect pocket.can_contain( it ) to fail, with an expected reason and contain_code
+static void expect_cannot_contain( item_pocket &pocket, item &it, std::string expect_reason,
+                                   item_pocket::contain_code expect_code )
+{
+    CAPTURE( it.tname() );
+    ret_val<item_pocket::contain_code> rate_can = pocket.can_contain( it );
+    CHECK_FALSE( rate_can.success() );
+    CHECK( rate_can.str() == expect_reason );
+    CHECK( rate_can.value() == expect_code );
+}
+
+// Call pocket.insert_item( it ) and expect it to be successful (contain_code::SUCCESS)
+static void expect_can_insert( item_pocket &pocket, item &it )
+{
+    CAPTURE( it.tname() );
+    ret_val<item_pocket::contain_code> rate_can = pocket.can_contain( it );
+    CHECK( rate_can.success() );
+    CHECK( rate_can.str().empty() );
+    CHECK( rate_can.value() == item_pocket::contain_code::SUCCESS );
+}
+
+// Call pocket.insert_item( it ) and expect it to fail, with an expected reason and contain_code
+static void expect_cannot_insert( item_pocket &pocket, item &it, std::string expect_reason,
+                                  item_pocket::contain_code expect_code )
```

Comment:
```suggestion
static void expect_cannot_insert( item_pocket &pocket, item &it, const std::string &expect_reason,
                                  item_pocket::contain_code expect_code )
```
passing a const reference to a string is much faster, though it doesn't matter for int and other primitive types, which an enum is